File: frontend/src/tools/purchase-records/schemas.ts
import { z } from "zod"

export const purchaseRecordsViewSchema = z.object({
    view: z
        .enum(["records", "categories", "subcategories"])
        .catch("records")
        .default("records"),
})

export const recordFilterSchema = z
    .object({
        page: z.number().int().min(1),
        page_size: z.number().int().min(1).max(100),
        title: z.string().trim().optional(),
        category_id: z.string().trim().optional(),
        subcategory_id: z.string().trim().optional(),
        purchase_date_start: z.string().trim().optional(),
        purchase_date_end: z.string().trim().optional(),
        owner_id: z.string().trim().optional(),
    })
    .refine(
        (data) => {
            if (!data.purchase_date_start || !data.purchase_date_end) {
                return true
            }

            return data.purchase_date_start <= data.purchase_date_end
        },
        {
            message: "开始日期不能晚于结束日期",
            path: ["purchase_date_end"],
        },
    )

export const purchaseRecordFormSchema = z.object({
    title: z.string().trim().min(1, { message: "标题不能为空" }),
    remark: z.string().trim(),
    amount: z
        .string()
        .trim()
        .min(1, { message: "金额不能为空" })
        .regex(/^(0|[1-9]\d*)(\.\d{1,2})?$/, {
            message: "金额最多保留两位小数",
        }),
    purchase_date: z.string().trim().min(1, { message: "购买日期不能为空" }),
    category_id: z.string().trim().min(1, { message: "大类不能为空" }),
    subcategory_id: z.string().trim(),
})

export const categoryFormSchema = z.object({
    name: z.string().trim().min(1, { message: "大类名称不能为空" }),
    is_active: z.boolean(),
})

export const subcategoryFormSchema = z.object({
    category_id: z.string().trim().min(1, { message: "大类不能为空" }),
    name: z.string().trim().min(1, { message: "小类名称不能为空" }),
    is_active: z.boolean(),
})

File: backend/app/modules/purchase_records/schemas.py
import uuid
from datetime import date
from decimal import Decimal
from typing import Any

from pydantic import field_serializer
from sqlmodel import Field, SQLModel


class RequestEnvelope(SQLModel):
    request_id: str | None = None
    ts: int | None = None
    payload: dict[str, Any] = Field(default_factory=dict)


class ErrorBody(SQLModel):
    details: dict[str, Any] = Field(default_factory=dict)


class ErrorResponse(SQLModel):
    version: str = "1.0"
    success: bool = False
    code: str
    message: str
    request_id: str
    ts: int
    error: ErrorBody


class PurchaseRecordPayload(SQLModel):
    title: str = Field(min_length=1)
    remark: str | None = None
    amount: Decimal
    purchase_date: date
    category_id: uuid.UUID
    subcategory_id: uuid.UUID | None = None


class PurchaseCategoryPayload(SQLModel):
    name: str = Field(min_length=1)
    is_active: bool


class PurchaseSubcategoryPayload(SQLModel):
    category_id: uuid.UUID
    name: str = Field(min_length=1)
    is_active: bool


class PurchaseRecordItem(SQLModel):
    id: uuid.UUID
    title: str
    remark: str | None = None
    amount: Decimal
    purchase_date: date
    category_id: uuid.UUID
    category_name: str
    subcategory_id: uuid.UUID | None = None
    subcategory_name: str | None = None
    owner_id: uuid.UUID
    owner_name: str

    @field_serializer("amount")
    def serialize_amount(self, value: Decimal) -> str:
        return f"{value:.2f}"


class PurchaseRecordListData(SQLModel):
    records: list[PurchaseRecordItem]
    total: int


class PurchaseRecordListResponse(SQLModel):
    version: str = "1.0"
    success: bool = True
    code: str = "OK"
    message: str = "success"
    request_id: str
    ts: int
    data: PurchaseRecordListData


class PurchaseRecordDetailResponse(SQLModel):
    version: str = "1.0"
    success: bool = True
    code: str = "OK"
    message: str = "success"
    request_id: str
    ts: int
    data: PurchaseRecordItem


class PurchaseRecordDeleteData(SQLModel):
    id: uuid.UUID
    is_deleted: bool
    deleted_at: str | None = None


class PurchaseRecordDeleteResponse(SQLModel):
    version: str = "1.0"
    success: bool = True
    code: str = "OK"
    message: str = "success"
    request_id: str
    ts: int
    data: PurchaseRecordDeleteData


class PurchaseCategoryItem(SQLModel):
    id: uuid.UUID
    name: str
    is_active: bool


class PurchaseCategoryListData(SQLModel):
    categories: list[PurchaseCategoryItem]


class PurchaseCategoryListResponse(SQLModel):
    version: str = "1.0"
    success: bool = True
    code: str = "OK"
    message: str = "success"
    request_id: str
    ts: int
    data: PurchaseCategoryListData


class PurchaseCategoryDetailResponse(SQLModel):
    version: str = "1.0"
    success: bool = True
    code: str = "OK"
    message: str = "success"
    request_id: str
    ts: int
    data: PurchaseCategoryItem


class PurchaseSubcategoryItem(SQLModel):
    id: uuid.UUID
    category_id: uuid.UUID
    category_name: str
    name: str
    is_active: bool


class PurchaseSubcategoryListData(SQLModel):
    subcategories: list[PurchaseSubcategoryItem]


class PurchaseSubcategoryListResponse(SQLModel):
    version: str = "1.0"
    success: bool = True
    code: str = "OK"
    message: str = "success"
    request_id: str
    ts: int
    data: PurchaseSubcategoryListData


class PurchaseSubcategoryDetailResponse(SQLModel):
    version: str = "1.0"
    success: bool = True
    code: str = "OK"
    message: str = "success"
    request_id: str
    ts: int
    data: PurchaseSubcategoryItem
